### PR TITLE
Added support for running DCOS-Diagnostics in the agent role on a Windows agent node

### PR DESCRIPTION
**Pull request by soccerGB**
https://jira.mesosphere.com/browse/DCOS_OSS-2151

This PR is to enable DCOS-Diagnostics to support agent role a Windows agent node inside a DC/OS cluster with Windows nodes

Changes in this PR include:

All the necessary changes needed for successful building Windows version of dcos-diagnostics.exe

Changes for making sure all unitests (mainly under api and runner subdirs) passed

Added logics for supporting /system/health/v1 endpoint with information collected/queried from interacting with Windows Service Control Manager on installed DC/OS services state

Other than the existing unittests, changes in the PR was tested on in the following scenario:
dcos-diagnostics.exe ran as an agent in daemon mode on a Windows agent node inside a hybrid DC/OS cluster with Linux and Windows agent nodes. In the DC/OS UI, the correct healthy states (expected healthy or unhealthy) was shown for each Windows agent node.

### CODE REVIEW

**Comment by soccerGB** [comment]
@mnaboka

Could you help review this PR to see if those changes meet the bar?
Any feedback will be greatly appreciated!

**Comment by soccerGB** [comment]
@branden 
Could you help review this PR to see if those changes meet the bar?
Any feedback will be greatly appreciated!

**Comment by mnaboka** [comment]
@soccerGB thanks for submitting this change list. I am in the process of reviewing this PR, meanwhile, could you explain how to run this on windows? i want to make sure it works

**Comment by mnaboka** [comment]
opened a DC/OS PR to run end to end integration tests https://github.com/dcos/dcos/pull/2569
EE DC/OS https://github.com/mesosphere/dcos-enterprise/pull/2400

File: api/diagnostics.go [hunk 1]
```diff
@@ -771,7 +773,9 @@ type CommandProvider struct {
 }
 
 func loadExternalProviders(cfg *config.Config) (externalProviders LogProviders, err error) {
+	logrus.Infof("loadExternalProviders:")
 	// return if cfg file not found.
+	logrus.Infof("cfg.FlagDiagnosticsBundleEndpointsConfigFile: %s", cfg.FlagDiagnosticsBundleEndpointsConfigFile)
```
Comment (mnaboka) [inline api/diagnostics.go]: please remove the Info message or turn them into debug

File: api/diagnostics.go [hunk 2]
```diff
@@ -786,10 +790,18 @@ func loadExternalProviders(cfg *config.Config) (externalProviders LogProviders,
 	if err = json.Unmarshal(endpointsConfig, &externalProviders); err != nil {
 		return externalProviders, err
 	}
+
+	// for debugging purpose
+	logrus.Infof("HTTPEndpoints.count: %d", len(externalProviders.HTTPEndpoints))
+	for _, endpoint := range externalProviders.HTTPEndpoints {
+		logrus.Infof("Endpoint: Port %d , endpoint url: %s FileName:%s", endpoint.Port, endpoint.URI, endpoint.FileName)
+	}
```
Comment (mnaboka) [inline api/diagnostics.go]: remove this block

File: api/diagnostics.go [hunk 3]
```diff
@@ -786,10 +790,18 @@ func loadExternalProviders(cfg *config.Config) (externalProviders LogProviders,
...
+
 	return externalProviders, nil
 }
 
 func loadInternalProviders(cfg *config.Config, DCOSTools DCOSHelper) (internalConfigProviders LogProviders, err error) {
+	logrus.Infof("loadInternalProviders:")
```
Comment (mnaboka) [inline api/diagnostics.go]: remove this info or switch to debug

File: api/diagnostics.go [hunk 4]
```diff
@@ -822,6 +834,12 @@ func loadInternalProviders(cfg *config.Config, DCOSTools DCOSHelper) (internalCo
 		FileName: "dcos-diagnostics-health.json",
 	})
 
+	logrus.Infof("role: %s  port: %d BaseRoute: %s", role, port, BaseRoute)
+	// for debugging purpose
+	for _, endpoint := range httpEndpoints {
+		logrus.Infof("Endpoint: Port %d , endpoint url: %s FileName:%s", endpoint.Port, endpoint.URI, endpoint.FileName)
+	}
+
```
Comment (mnaboka) [inline api/diagnostics.go]: remove this entire block

File: api/diagnostics_test.go [hunk 5]
```diff
@@ -386,7 +386,7 @@ func (s *DiagnosticsTestSuit) TestRunSnapshot() {
 	s.assert.Equal(responseJSON.Status, "Job has been successfully started")
 	s.assert.NotEmpty(responseJSON.Extra.LastBundleFile)
 	time.Sleep(2 * time.Second)
-	snapshotFiles, err := ioutil.ReadDir("/tmp/snapshot-test")
+	snapshotFiles, err := ioutil.ReadDir(DiagnosticsBundleDir)
 	s.assert.NoError(err)
 	s.assert.True(len(snapshotFiles) > 0)
```
Comment (mnaboka) [inline api/diagnostics_test.go]: any change to `*_test.go` should be a separate commit.

File: api/handlers_test.go [hunk 6]
```diff
@@ -7,22 +7,31 @@ import (
...
 )
 
+var DiagnosticsBundleDir = "/tmp/snapshot-test"
 var testCfg *config.Config
 
 func init() {
+	if runtime.GOOS == "windows" {
+		DiagnosticsBundleDir = os.Getenv("SYSTEMDRIVE") + "\\tmp\\snapshot-test"
+	}
+	logrus.Infof("DiagnosticsBundleDir=%v", DiagnosticsBundleDir)
```
Comment (mnaboka) [inline api/handlers_test.go]: ideally we should not be using logrus in tests

File: api/router.go [hunk 7]
```diff
@@ -318,6 +318,10 @@ func getRoutes(debug bool) []routeHandler {
 		}...)
 	}
 
+	for i, route := range routes {
+		logrus.Debug(i, ": route.url=", route.url)
+	}
+
```
Comment (mnaboka) [inline api/router.go]: remove this block

File: cmd/check.go [hunk 8]
```diff
@@ -96,6 +96,13 @@ var checkCmd = &cobra.Command{
 
 func init() {
 	RootCmd.AddCommand(checkCmd)
+
+	if runtime.GOOS == "windows" {
+		defaultRunnerConfig = os.Getenv("SYSTEMDRIVE") + 
+		                       "\\DCOS\\diagnostics\\config\\dcos-diagnostics-runner-config.json"
+	}
```
Comment (mnaboka) [inline cmd/check.go]: maybe use a constant, similar to `defaultRunnerConfig` to define a default path for the config on windows machine?

File: cmd/daemon.go [hunk 9]
```diff
@@ -129,6 +130,13 @@ func startDiagnosticsDaemon() {
 	if defaultConfig.FlagForceTLS {
 		options = append(options, nodeutil.OptionMesosStateURL(defaultStateURL.String()))
 	}
+	if runtime.GOOS == "windows" {
+		detectIPScriptPath := os.Getenv("SYSTEMDRIVE") + "\\DCOS\\diagnostics\\detect_ip.ps1"
+		options = append(options, nodeutil.OptionDetectIP(detectIPScriptPath))
```
Comment (mnaboka) [inline cmd/daemon.go]: same here, can we use const or variable to define a default path on windows machine?

File: cmd/listener_windows.go [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+	"os"
+)
+
+func getListener(unsetEnv bool) ([]net.Listener, error) {
+	return nil, nil
+}
+
+func getFiles(unsetEnv bool) []*os.File {
+	return nil
+}
```
Comment (mnaboka) [inline cmd/listener_windows.go]: maybe instead of having 2 separate files like `listener_linux.go` and `listener_windows.go` we should define a simple `if linux then... else if windows then ...` ?

File: api/diagnostics.go (hunk 2 above)
Comment (soccerGB) [reply api/diagnostics.go]: @mnaboka 

Here is how I tested this dcos-diagnostics on window

https://github.com/soccerGB/Docs/blob/master/dcos/diagnosticsservice/HowToRunDiagnosticsOnWindows.md

File: api/diagnostics.go [hunk 11]
```diff
@@ -126,7 +126,7 @@ func (j *DiagnosticsJob) run(req bundleCreateRequest, dt *Dt) (createResponse, e
 	_, err = os.Stat(dt.Cfg.FlagDiagnosticsBundleDir)
 	if os.IsNotExist(err) {
 		logrus.Infof("Directory: %s not found, attempting to create one", dt.Cfg.FlagDiagnosticsBundleDir)
-		if err := os.Mkdir(dt.Cfg.FlagDiagnosticsBundleDir, os.ModePerm); err != nil {
+		if err := os.MkdirAll(dt.Cfg.FlagDiagnosticsBundleDir, os.ModePerm); err != nil {
```
[the remark this answers is not in the document]
Comment (soccerGB) [reply api/diagnostics.go]: On linux, the default value for  FlagDiagnosticsBundleDir is "/tmp/snapshot-test".  /tmp is a directory that exists upon the Linux OS boot up for virtually all the Linux distros. So, calling Mkdir works fine.
 
On Windows, \tmp\snapshot-test is used as the default value for making it as similar to the code base. However,  tmp is not a pre-created directory on a clean windows OS. Replacing MkdirAll with Mkdir would make the directory not found (tmp) directory go way.

File: api/diagnostics.go [hunk 12]
```diff
@@ -272,6 +272,7 @@ func (j *DiagnosticsJob) runBackgroundJob(nodes []Node, cfg *config.Config, DCOS
 		updateSummaryReport("START collecting logs", node, "", summaryReport)
 		url := fmt.Sprintf("http://%s:%d%s/logs", node.IP, port, BaseRoute)
 		endpoints := make(map[string]string)
+		logrus.Errorf("url: %s", url)
```
[the remark this answers is not in the document]
Comment (soccerGB) [reply api/diagnostics.go]: Good catch. I will remove it.

File: api/diagnostics.go [hunk 13]
```diff
@@ -281,6 +282,7 @@ func (j *DiagnosticsJob) runBackgroundJob(nodes []Node, cfg *config.Config, DCOS
 			j.JobProgressPercentage += percentPerNode
 			continue
 		}
+		logrus.Infof("json.Unmarshal: %v", body)
```
[the remark this answers is not in the document]
Comment (soccerGB) [reply api/diagnostics.go]: Oops. I will remove it.

File: api/diagnostics.go (hunk 1 above)
Comment (soccerGB) [reply api/diagnostics.go]: I will replace it with Debugf

File: api/diagnostics.go (hunk 2 above)
Comment (soccerGB) [reply api/diagnostics.go]: I will remove this debug only block

File: api/router.go (hunk 7 above)
Comment (soccerGB) [reply api/router.go]: I will remove this debug only block

File: api/diagnostics.go (hunk 3 above)
Comment (soccerGB) [reply api/diagnostics.go]: Will remove.

File: api/diagnostics.go (hunk 4 above)
Comment (soccerGB) [reply api/diagnostics.go]: Will remove.

File: api/handlers_test.go (hunk 6 above)
Comment (soccerGB) [reply api/handlers_test.go]: That was the only message that's using logrus and it's not an essential message. I will remove them from this test file to make it cleaner.

File: cmd/check.go (hunk 8 above)
Comment (soccerGB) [reply cmd/check.go]: Good idea.

I will add that.

File: cmd/daemon.go (hunk 9 above)
Comment (soccerGB) [reply cmd/daemon.go]: good idea.  I will replace it with a const.

File: cmd/listener_windows.go (hunk 10 above)
Comment (soccerGB) [reply cmd/listener_windows.go]: That was what I tried in the first place. It won't compile because the go-systemd/activation package only exists in the Linux version of the Golang.

File: api/diagnostics_test.go (hunk 5 above)
Comment (soccerGB) [reply api/diagnostics_test.go]: Let me see what i can do here....

**Comment by soccerGB** [comment]
@mnaboka, thanks for your great feedbacks
Per your suggestions, I broke the original commit into several smaller commits and also addressed your previous review feedbacks

**Comment by mnaboka** [comment]
@soccerGB thank you for you effort! this is a very big and great job porting diagnostics service to windows! I will try to setup windows CI tomorrow, to make sure we are able to run at least the unittests in our CI and when this is done, we will merge you PR. Congrats! 🎉 

Also, would you rebase your change to the latest master?

**Comment by mnaboka** [comment]
@soccerGB also, OSS and EE integration tests passed! let's rebase this PR to the latest master and i'll rerun the tests one more time, just to be sure.

**Comment by soccerGB** [comment]
@mnaboka, its rebased to the latest master.

**Comment by mnaboka** [comment]
🚢
